src/tcp_tunnel_tools.js

```diff
@@ -1,78 +1,66 @@
 const net = require('net');
-const TcpTunnel = require('./tcp_tunnel');
+const { chain } = require('./chain');
 const { nodeify } = require('./utils/nodeify');
 
 const runningServers = {};
 
-function createTunnel(proxyUrl, targetHost, providedOptions = {}, callback) {
+function createTunnel(proxyUrl, targetHost, options, callback) {
     const parsedProxyUrl = new URL(proxyUrl);
     if (parsedProxyUrl.protocol !== 'http:') {
         throw new Error(`The proxy URL must have the "http" protocol (was "${proxyUrl}")`);
     }
 
-    // TODO: More and better validations - yeah, make sure targetHost is really a hostname
-    const [trgHostname, trgPort] = (targetHost || '').split(':');
-    if (!trgHostname || !trgPort) throw new Error('The target host needs to include both hostname and port.');
+    const url = new URL(`connect://${targetHost || ''}`);
 
-    const options = {
-        verbose: false,
-        hostname: 'localhost',
-        port: null,
-        ...providedOptions,
-    };
+    if (!url.hostname) {
+        throw new Error('Missing target hostname');
+    }
+
+    if (!url.port) {
+        throw new Error('Missing target port');
+    }
+
+    const verbose = options && options.verbose;
 
     const server = net.createServer();
+    let address;
 
     const log = (...args) => {
-        if (options.verbose) console.log(...args);
+        if (verbose) console.log(...args);
     };
 
+    server.log = log;
+
     server.on('connection', (srcSocket) => {
-        const { port } = server.address();
+        runningServers[address].connections.push(srcSocket);
 
-        runningServers[port].connections.push(srcSocket);
         const remoteAddress = `${srcSocket.remoteAddress}:${srcSocket.remotePort}`;
-        log('new client connection from %s', remoteAddress);
-
-        srcSocket.pause();
-
-        const tunnel = new TcpTunnel({
-            srcSocket,
-            upstreamProxyUrlParsed: parsedProxyUrl,
-            trgParsed: {
-                hostname: trgHostname,
-                port: trgPort,
-            },
-            log,
+        log(`new client connection from ${remoteAddress}`);
+
+        chain({
+            request: { url: targetHost },
+            source: srcSocket,
+            handlerOpts: { upstreamProxyUrlParsed: parsedProxyUrl },
+            server,
+            plain: true,
         });
+    });
 
-        tunnel.run();
+    const promise = new Promise((resolve, reject) => {
+        server.once('error', reject);
 
-        srcSocket.on('data', onConnData);
-        srcSocket.on('close', onConnClose);
-        srcSocket.on('error', onConnError);
```

Comment:
Are we removing all the logs? Why?

Reply:
Ah, sorry, forgot about them by accident. Will restore this.

Reply:
`error` is logged inside `chain`. `data` makes the logs too messy. I'll just log on `close`.

Reply:
Fixed.